docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -419,6 +419,73 @@ public class TestRecorder {
 ----
 <1> Pass a contextual reference of `Bar` to the constructor of `Foo`.
 
+=== Inactive Synthetic Beans
+
+In the case when one needs to register multiple synthetic beans at build time but only wants a subset of them active at runtime, it is useful to be able to mark a synthetic bean as _inactive_.
+This is done by configuring a "check active" procedure, which should be a `Supplier<ActiveResult>` obtained from a recorder:
+
+.Inactive Synthetic Bean - Build Step Example
+[source,java]
+----
+@BuildStep
+@Record(RUNTIME_INIT)
+SyntheticBeanBuildItem syntheticBean(TestRecorder recorder) {
+    return SyntheticBeanBuildItem.configure(Foo.class)
+            .scope(Singleton.class)
+            .startup() // <1>
+            .checkActive(recorder.isFooActive()) // <2>
+            .createWith(recorder.createFoo())
+            .done();
+}
+----
+<1> A bean that might be inactive is typically initialized eagerly, to make sure that an error is thrown at application startup.
+    If the bean is in fact inactive, but is not injected into an always-active bean, eager initialization is skipped and no error is thrown.
+<2> Configures the "check active" procedure.
+
+.Inactive Synthetic Bean - Recorder Example
+[source,java]
+----
+@Recorder
+public class TestRecorder {
+    public Supplier<ActiveResult> isFooActive() {
+        return () -> {
+            if (... should not be active ...) { // <1>
+                return ActiveResult.inactive("explanation"); // <2>
+            }
+            return ActiveResult.active();
+        };
+    }
+
+    public Function<SyntheticCreationalContext<Foo>, Foo> createFoo() {
+        return (context) -> {
+            return new Foo();
+        };
+    }
+}
+----
+<1> The condition when the synthetic bean should be inactive.
+<2> Proper explanation of why the bean is inactive.
+    Another inactive `ActiveResult` may also be provided as a cause, if this bean's inactivity stems from another bean's inactivity.
+
+If an inactive bean is injected somewhere, or is dynamically looked up, an `InactiveBeanException` is thrown.
+The error message contains the reason (from the `ActiveResult`), the cause chain (also from the `ActiveResult`), and possibly also a list of all injection points that resolve to this bean.
+
+This means that a possibly inactive bean should only be injected using `Instance<\...>`.
+Further, the bean should be checked before obtaining the actual instance:
+
+[source,java]
+----
+import io.quarkus.arc.InjectableInstance;
+
+@Inject
+InjectableInstance<Foo> foo;
+
+if (foo.getHandle().getBean().isActive()) {
```

Comment:
Could we perhaps make our impl of `Instance.isResolvable` aware of this and use that instead?
User would then not need to know the gritty details of bean being possibly inactive and would just use standard APIs.

Reply:
Interesting idea. Note that we already filter beans that are [suppressed](https://quarkus.io/guides/cdi-reference#declaratively-choose-beans-that-can-be-obtained-by-programmatic-lookup). Maybe an inactive bean should be automatically suppressed?

Reply:
That is a good point!

Reply:
I need a way to find beans that are inactive, in order to propagate their "inactive reason" to other beans that also become inactive as a result. See [this](https://github.com/quarkusio/quarkus/pull/41929/files#diff-ea1e7589789d9a9f8ee570a12d6cd973cb23cd7aa97f4a14eea1266a110779ccR59-R84) for example:

```java
    public Supplier<ActiveResult> flywayCheckActiveSupplier(String dataSourceName) {
        return new Supplier<ActiveResult>() {
            @Override
            public ActiveResult get() {
                // Flyway beans are inactive when the datasource itself is inactive.
                var dataSourceBean = AgroalDataSourceUtil.dataSourceInstance(dataSourceName).getHandle().getBean();
                var dataSourceActive = dataSourceBean.checkActive();
                if (!dataSourceActive.value()) {
                    return ActiveResult.inactive(
                            String.format(Locale.ROOT,
                                    "Flyway for datasource '%s' was deactivated automatically because this datasource was deactivated.",
                                    dataSourceName),
                            dataSourceActive);
                }

                // Note: When quarkus.flyway.active is set to false, Flyway beans are still available.
                //       The property only controls automatic execution on startup.
                // TODO should we change quarkus.flyway.active (see ^) to align on other extensions?
                //   See https://github.com/quarkusio/quarkus/issues/42244.
                //   We'd have something like quarkus.liquibase.startup.enabled controlling startup behavior,
                //   and *if necessary* quarkus.liquibase.active controlling bean availability
                //   (though IMO controlling that at the datasource level would be enough).
                return ActiveResult.active();
            }
        };
    }
```

If you just filter out inactive beans, will I still be able to find them and get their "inactive reason"?

Or do you want to implement the "inactive bean propagation" directly in Arc? That's tempting, but I can see it having unintended side-effects in the future (deactivating beans where it makes no sense).

Reply:
> If you just filter out inactive beans, will I still be able to find them and get their "inactive reason"?

My initial comment didn't mean to filter beans - we could just alter how `Instance.isResolvable()` work to account for inactive beans (i.e. returning false if it is inactive). You'd still be able to get the metadata.

Either way, with https://github.com/quarkusio/quarkus/pull/41810#discussion_r1789647409 in mind, I am not sure it's worth the hassle :)

Reply:
So I don't think we can make inactive beans suppressed, because those cannot be obtained via `Instance` at all.

Whether we can make the `Instance` non-resolvable, I'm not sure. Let me think about that.